Trace frames with `numpy.ndarray`.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109715

Fix: #109604

This PR enables tracing of frames that contain NumPy arrays (without PyTorch modules or
tensors).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ysiraichi committed:
Trace frames with `numpy.ndarray`.

Fix: #109604

This PR enables tracing of frames that contain NumPy arrays (without PyTorch modules or
tensors).

[ghstack-poisoned]

lezcano commented on torch/_dynamo/convert_frame.py:
```diff
@@ -2,6 +2,7 @@
 import functools
 import itertools
 import logging
+import numpy
```
You need to conditionally import it. See the pattern in other `_dynamo` files.

lezcano commented on torch/_dynamo/convert_frame.py:
```diff
@@ -177,7 +178,7 @@ def has_tensor(obj):
         seen_ids[obj_id] = False
 
-        if isinstance(obj, (torch.Tensor, torch.nn.Module)) or (
+        if isinstance(obj, (torch.Tensor, torch.nn.Module, numpy.ndarray)) or (
```
Just add that if `conf.trace_numpy is True`.

ysiraichi committed:
Update on "Trace frames with `numpy.ndarray`."


Fix: #109604

This PR enables tracing of frames that contain NumPy arrays (without PyTorch modules or
tensors).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "Trace frames with `numpy.ndarray`."


Fix: #109604

This PR enables tracing of frames that contain NumPy arrays (without PyTorch modules or
tensors).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano commented on test/dynamo/test_misc.py:
```diff
@@ -7074,6 +7075,23 @@ def fn(it):
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    @torch._dynamo.config(trace_numpy=True)
+    def test_trace_ndarray_frame(self):
+        import numpy as np
```
NumPy is already imported globally. It's fine to import NumPy globally in tests, but not in the `torch/` folder.

lezcano commented on test/dynamo/test_misc.py:
```diff
@@ -7074,6 +7075,23 @@ def fn(it):
         self.assertEqual(list(eager), list(compiled))
         self.assertEqual(counter.frame_count, 1)
 
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    @torch._dynamo.config(trace_numpy=True)
```
I believe these two are not necessary really. The numpy import at the top would break before the first one is hit (it's fine, numpy is a required dependency during testing I believe), and the second one is already the default value. 

No one has complained about the naked nmpy imports in the tests. If they do, we'll need to fix it in many other parts of the code as well

ysiraichi committed:
Update on "Trace frames with `numpy.ndarray`."


Fix: #109604

This PR enables tracing of frames that contain NumPy arrays (without PyTorch modules or
tensors).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano reviewed (approved):
Thank you!

lezcano commented:
@pytorchbot merge

ysiraichi commented:
CI failures seem legit. Apparently, this is not as easy as I thought.

lezcano commented:
Could it be that, since we're tracing more now, we're uncovering new issues?

lezcano commented:
In particular it's looking int
https://github.com/mindee/doctr/blob/28e5375c399ae5dd146a1b7e13bd5831acb6dc2d/doctr/models/detection/core.py#L84-L90

There's the list comprehension with dynamic shape, so there's a graph break there, but then it continues trying to trace through
```python
                cv2.morphologyEx(bmap[..., idx], cv2.MORPH_OPEN, self._opening_kernel)
```
and chokes saying that the input is not an ndarray? Perhaps this is starting tracing again, but it's somehow not calling back the function that reconstructs the arrays from the tensors? We usually do that via the `to_numpy_helper` within codegen.py, in case this is relevant.

ysiraichi commented:
Yeah. That's one thing. There's also `test_logaddexp_cpu_complex128`, more specifically `scipy.special.logsumexp`, which gives different results when traced by dynamo. In summary, dynamo can't trace it completely because of `numpy.ma.isMaskedArray` (graph-break). Then, for some reason, the result is different.

lezcano commented:
Why do we try to trace through scipy functions, rather than graphbreaking?

ysiraichi commented:
There's just nothing stopping dynamo from doing that. I will try adding a graph-break there.

lezcano commented:
Fair enough. I guess that we should figure out why these two are not graphbreaking properly like, why is the first one not converting the tensor back into an array, and in the second one, why are we not breaking properly on maskedarrays.

lezcano commented:
being landed at https://github.com/pytorch/pytorch/pull/110512
